app/uk/gov/hmrc/mtdsatestsupportapi/controllers/requestParsers/validators/CreateBusinessValidator.scala

```diff
@@ -0,0 +1,117 @@
+import play.api.libs.json.{JsDefined, JsLookupResult, JsValue}
+import uk.gov.hmrc.mtdsatestsupportapi.controllers.requestParsers.validators.validations._
+import uk.gov.hmrc.mtdsatestsupportapi.models.request.createBusiness.{Business, CreateBusinessRawData}
+
+import java.time.{Clock, LocalDate}
+import javax.inject.{Inject, Singleton}
+import scala.math.Ordered.orderingToOrdered
+import scala.util.{Failure, Success, Try}
+
+@Singleton
+class CreateBusinessValidator @Inject() (clock: Clock) extends Validator[CreateBusinessRawData] {
+
+  override def validate(data: CreateBusinessRawData): List[MtdError] =
+    (for {
+      _ <- parameterFormatValidation(data)
+      _ <- bodyEnumValidation(data.body)
+      _ <- bodyFormatValidation(data)
+      _ <- bodyFieldValidation(data)
+    } yield ()).swap.getOrElse(Nil)
+
+  private def parameterFormatValidation(data: CreateBusinessRawData): Either[List[MtdError], Unit] = {
+    val errors = NinoValidation.validate(data.nino)
+
+    errorsResult(errors)
+  }
+
+  private def bodyEnumValidation(body: JsValue): Either[List[MtdError], Unit] = {
+    val errors = validateAccountingType(body) ++ validateTypeOfBusiness(body) ++
+      validateLatencyIndicator(body, "/latencyDetails/latencyIndicator1") ++
+      validateLatencyIndicator(body, "/latencyDetails/latencyIndicator2")
+
+    errorsResult(errors)
+  }
+
+  private def bodyFormatValidation(data: CreateBusinessRawData): Either[List[MtdError], Unit] = {
+    val errors = JsonFormatValidation.validate[Business](data.body) match {
+      case Nil          => NoValidationErrors
+      case schemaErrors => schemaErrors
+    }
+
+    errorsResult(errors)
+  }
+
+  private def bodyFieldValidation(data: CreateBusinessRawData): Either[List[MtdError], Unit] = {
+    val business = data.body.as[Business]
+
+    val errors =
+      business.businessAddressCountryCode.map(CountryCodeValidation.validate(_)).getOrElse(Nil) ++
+        validatePostcode(business) ++
+        validateMissingPostcode(business) ++
+        business.firstAccountingPeriodStartDate.map(validateDate(_, "/firstAccountingPeriodStartDate")).getOrElse(Nil) ++
+        business.firstAccountingPeriodEndDate.map(validateDate(_, "/firstAccountingPeriodEndDate")).getOrElse(Nil) ++
+        business.latencyDetails.map(ld => validateDate(ld.latencyEndDate, "/latencyDetails/latencyEndDate")).getOrElse(Nil) ++
+        business.commencementDate.map(validateCommencementDate).getOrElse(Nil) ++
+        business.cessationDate.map(validateDate(_, "/cessationDate")).getOrElse(Nil)
+
+    errorsResult(errors)
+  }
+
+  private def validateDate(value: String, path: String): Seq[MtdError] =
+    DateValidation.validate(value, DateFormatError.withExtraPath(path))
+
+  private def validateAccountingType(json: JsValue): Seq[MtdError] =
+    (json \ "accountingType").asOpt[String].map(AccountingTypeValidation.validate(_)).getOrElse(Nil)
+
+  private def validateTypeOfBusiness(json: JsValue): Seq[MtdError] =
```

Comment:
`typeOfBusiness` is mandatory, not optional. Was asOpt here just used for convenience to apply the `TypeOfBusinessValidation` validation to it?

Reply:
It's treated as optional here because the `bodyEnumValidation` needs to be done before the `bodyFormatValidation` (which will otherwise fail less nicely when trying to deserialise invalid business types). But doing it before `bodyFormatValidation` we don't know whether fields that are mandatory are actually present.